File: backend/app/api/routes/sessions.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from typing import Optional

from app.core.msf_client import msf_client
from app.api.routes.auth import get_current_active_user

router = APIRouter()


class SessionCommand(BaseModel):
    command: str


class SessionResponse(BaseModel):
    output: str


@router.get("")
async def list_sessions(user=Depends(get_current_active_user)):
    """List all active sessions."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        sessions = await msf_client.list_sessions()
        # Transform to a more usable format
        result = []
        for sid, info in sessions.items():
            result.append({
                "id": int(sid),
                "type": info.get("type", "unknown"),
                "tunnel_local": info.get("tunnel_local", ""),
                "tunnel_peer": info.get("tunnel_peer", ""),
                "via_exploit": info.get("via_exploit", ""),
                "via_payload": info.get("via_payload", ""),
                "desc": info.get("desc", ""),
                "info": info.get("info", ""),
                "workspace": info.get("workspace", ""),
                "session_host": info.get("session_host", ""),
                "session_port": info.get("session_port", 0),
                "target_host": info.get("target_host", ""),
                "username": info.get("username", ""),
                "uuid": info.get("uuid", ""),
                "exploit_uuid": info.get("exploit_uuid", ""),
                "routes": info.get("routes", []),
                "arch": info.get("arch", ""),
                "platform": info.get("platform", ""),
            })
        return {"sessions": result, "count": len(result)}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{session_id}")
async def get_session(session_id: int, user=Depends(get_current_active_user)):
    """Get detailed information about a specific session."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        session = await msf_client.get_session(session_id)
        if not session:
            raise HTTPException(status_code=404, detail="Session not found")
        session["id"] = session_id
        return session
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/{session_id}/shell/write")
async def shell_write(session_id: int, cmd: SessionCommand, user=Depends(get_current_active_user)):
    """Send a command to a shell session."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        result = await msf_client.session_shell_write(session_id, cmd.command + "\n")
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{session_id}/shell/read")
async def shell_read(session_id: int, user=Depends(get_current_active_user)):
    """Read output from a shell session."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        output = await msf_client.session_shell_read(session_id)
        return {"output": output}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/{session_id}/meterpreter/write")
async def meterpreter_write(session_id: int, cmd: SessionCommand, user=Depends(get_current_active_user)):
    """Send a command to a meterpreter session."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        result = await msf_client.session_meterpreter_write(session_id, cmd.command)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{session_id}/meterpreter/read")
async def meterpreter_read(session_id: int, user=Depends(get_current_active_user)):
    """Read output from a meterpreter session."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        output = await msf_client.session_meterpreter_read(session_id)
        return {"output": output}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/{session_id}/meterpreter/run")
async def meterpreter_run(session_id: int, cmd: SessionCommand, user=Depends(get_current_active_user)):
    """Run a single meterpreter command and get output."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        output = await msf_client.session_meterpreter_run_single(session_id, cmd.command)
        return {"output": output}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{session_id}")
async def kill_session(session_id: int, user=Depends(get_current_active_user)):
    """Kill a session."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        result = await msf_client.kill_session(session_id)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: backend_ruby/standalone/routes/sessions.rb
# frozen_string_literal: true

require 'sinatra/base'

module MsfGui
  module Routes
    class Sessions < Sinatra::Base
      helpers Helpers::Response
      helpers Helpers::Validators

      configure do
        set :show_exceptions, false
      end

      # List all sessions
      get '/' do
        sessions = MsfGui.msf_client.list_sessions

        # Add session IDs to each session object
        result = sessions.map do |id, session|
          session.merge(id: id.to_s)
        end

        json_response(result)
      end

      # Get specific session
      get '/:session_id' do
        session = MsfGui.msf_client.get_session(params[:session_id])

        unless session
          json_error('Session not found', status: 404)
        end

        json_response(session.merge(id: params[:session_id]))
      end

      # Shell - write command
      post '/:session_id/shell/write' do
        body = parse_json_body
        command = body[:command] || body['command']

        unless command
          json_error('Command is required', status: 422)
        end

        result = MsfGui.msf_client.session_shell_write(params[:session_id], command)
        json_response({ success: true, write_count: result['write_count'] })
      end

      # Shell - read output
      get '/:session_id/shell/read' do
        result = MsfGui.msf_client.session_shell_read(params[:session_id])
        json_response({
          data: result['data'],
          seq: result['seq']
        })
      end

      # Meterpreter - write command
      post '/:session_id/meterpreter/write' do
        body = parse_json_body
        command = body[:command] || body['command']

        unless command
          json_error('Command is required', status: 422)
        end

        result = MsfGui.msf_client.session_meterpreter_write(params[:session_id], command)
        json_response({ success: true, result: result })
      end

      # Meterpreter - read output
      get '/:session_id/meterpreter/read' do
        result = MsfGui.msf_client.session_meterpreter_read(params[:session_id])
        json_response({
          data: result['data']
        })
      end

      # Meterpreter - run command and get output
      post '/:session_id/meterpreter/run' do
        body = parse_json_body
        command = body[:command] || body['command']

        unless command
          json_error('Command is required', status: 422)
        end

        result = MsfGui.msf_client.session_meterpreter_run_single(params[:session_id], command)
        json_response({
          data: result['data'] || result
        })
      end

      # Kill session
      delete '/:session_id' do
        result = MsfGui.msf_client.kill_session(params[:session_id])
        json_response({
          success: true,
          message: 'Session terminated',
          result: result
        })
      end

      # Upgrade shell to meterpreter
      post '/:session_id/upgrade' do
        body = parse_json_body
        lhost = body[:lhost] || body['lhost']
        lport = body[:lport] || body['lport'] || 4433

        unless lhost
          json_error('LHOST is required', status: 422)
        end

        # Use post/multi/manage/shell_to_meterpreter
        result = MsfGui.msf_client.run_module('post', 'multi/manage/shell_to_meterpreter', {
          'SESSION' => params[:session_id],
          'LHOST' => lhost,
          'LPORT' => lport.to_s
        })

        json_response({
          success: true,
          message: 'Upgrade initiated',
          job_id: result['job_id']
        })
      end

      # Run post module on session
      post '/:session_id/run' do
        body = parse_json_body
        module_name = body[:module] || body['module']
        options = body[:options] || body['options'] || {}

        unless module_name
          json_error('Module name is required', status: 422)
        end

        options['SESSION'] = params[:session_id]
        result = MsfGui.msf_client.run_module('post', module_name, options)

        json_response({
          success: true,
          job_id: result['job_id'],
          uuid: result['uuid']
        })
      end
    end
  end
end
